usim_pytest/test_usimpy/test_core.py

```diff
@@ -0,0 +1,79 @@
+import pytest
+
+from usim import Scope, time
+from usim.py import Environment
+from usim.py.exceptions import NotEmulatedError
+
+from ..utility import via_usim
+
+
+class TestEnvironment:
+    def test_events_done(self, env):
+        """env.run() stops automatically when all events are done"""
+        # side-effect to ensure something *is* run
+        def observable(env, log: list):
+            log.append(1)
+            yield env.timeout(1)
+            log.append(2)
+            yield env.timeout(1)
+
+        output = []
+        env.process(observable(env, output))
+        env.run()
+        assert output == [1, 2]
+        assert env.now == 2
+
+    def test_until_past(self, env):
+        """env.run(until) argument must not be in the past"""
+        with pytest.raises(ValueError):
+            env.run(-5)
```

Comment:
Consider also testing edge cases, such as `0`. I think I put a comment somewhere, that `0` should not raise a `ValueError`.